CI UpdateShardAllocationSettingsIT.testUpdateSameHostSetting failure

The `UpdateShardAllocationSettingsIT.testUpdateSameHostSetting` test failed in:
https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-unix-compatibility/os=fedora-29&&immutable/226/console

It did not reproduce locally with the command:
```
./gradlew :server:integTest \
--tests "org.elasticsearch.cluster.routing.allocation.decider.UpdateShardAllocationSettingsIT.testUpdateSameHostSetting" \
-Dtests.seed=58E6AE995D30C037 \
-Dtests.security.manager=true \
-Dtests.locale=zh-SG \
-Dtests.timezone=SystemV/CST6 \
-Dcompiler.java=12 \
-Druntime.java=11
```

It failed with:
```
java.lang.IllegalStateException: Some shards are still open after the threadpool terminated. Something is leaking index readers or store references.
	at __randomizedtesting.SeedInfo.seed([58E6AE995D30C037:1612DD2EB1BB20A2]:0)
	at org.elasticsearch.node.Node.awaitClose(Node.java:895)
	at org.elasticsearch.test.InternalTestCluster$NodeAndClient.close(InternalTestCluster.java:948)
	at org.elasticsearch.core.internal.io.IOUtils.close(IOUtils.java:104)
	at org.elasticsearch.core.internal.io.IOUtils.close(IOUtils.java:86)
	at org.elasticsearch.test.InternalTestCluster.close(InternalTestCluster.java:803)
	at org.elasticsearch.test.ESIntegTestCase.afterInternal(ESIntegTestCase.java:549)
	at org.elasticsearch.test.ESIntegTestCase.cleanUpCluster(ESIntegTestCase.java:1966)
```
and
```
1> [2019-08-26T00:19:50,844][INFO ][o.e.c.m.MetaDataDeleteIndexService] [node_t0] [idx/Ppy6SgGdRsqN9GUNjGB7NQ] deleting index
  1> [2019-08-26T00:19:50,890][WARN ][o.e.i.IndicesService     ] [node_t1] [idx/Ppy6SgGdRsqN9GUNjGB7NQ] failed to delete index
  1> org.elasticsearch.env.ShardLockObtainFailedException: [idx][0]: obtaining shard lock timed out after 0ms, previous lock details: [shard creation] trying to lock for [deleting index directory]
  1> 	at org.elasticsearch.env.NodeEnvironment$InternalShardLock.acquire(NodeEnvironment.java:860) ~[main/:?]
  1> 	at org.elasticsearch.env.NodeEnvironment.shardLock(NodeEnvironment.java:775) ~[main/:?]
  1> 	at org.elasticsearch.env.NodeEnvironment.lockAllForIndex(NodeEnvironment.java:718) ~[main/:?]
  1> 	at org.elasticsearch.env.NodeEnvironment.deleteIndexDirectorySafe(NodeEnvironment.java:670) ~[main/:?]
  1> 	at org.elasticsearch.indices.IndicesService.deleteIndexStoreIfDeletionAllowed(IndicesService.java:800) [main/:?]
  1> 	at org.elasticsearch.indices.IndicesService.deleteIndexStore(IndicesService.java:787) [main/:?]
  1> 	at org.elasticsearch.indices.IndicesService.removeIndex(IndicesService.java:691) [main/:?]
  1> 	at org.elasticsearch.indices.cluster.IndicesClusterStateService.deleteIndices(IndicesClusterStateService.java:375) [main/:?]
  1> 	at org.elasticsearch.indices.cluster.IndicesClusterStateService.applyClusterState(IndicesClusterStateService.java:241) [main/:?]
  1> 	at org.elasticsearch.cluster.service.ClusterApplierService.lambda$callClusterStateAppliers$5(ClusterApplierService.java:508) [main/:?]
  1> 	at java.lang.Iterable.forEach(Iterable.java:75) [?:?]
  1> 	at org.elasticsearch.cluster.service.ClusterApplierService.callClusterStateAppliers(ClusterApplierService.java:505) [main/:?]
  1> 	at org.elasticsearch.cluster.service.ClusterApplierService.applyChanges(ClusterApplierService.java:482) [main/:?]
  1> 	at org.elasticsearch.cluster.service.ClusterApplierService.runTask(ClusterApplierService.java:432) [main/:?]
  1> 	at org.elasticsearch.cluster.service.ClusterApplierService$UpdateTask.run(ClusterApplierService.java:176) [main/:?]
  1> 	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:699) [main/:?]
  1> 	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:252) [main/:?]
  1> 	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:215) [main/:?]
  1> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
  1> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
```

The `ShardLockObtainFailedException` quoted above isn't obviously correlated with the failure - at least, I see passing tests that emit this warning. I think that's ok, we can't always delete the shard straight away as there might still be ongoing searches, and we handle this here with a very generous timeout:

https://github.com/elastic/elasticsearch/blob/47fd1f20aa4de3ff353360a11412683862ab4f71/server/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java#L406-L410

However, it looks like this is failing with a `java.lang.InterruptedException` as the node is closing:

```
1> [2019-08-26T00:20:00,953][WARN ][o.e.i.c.IndicesClusterStateService] [node_t0] [[idx/Ppy6SgGdRsqN9GUNjGB7NQ]] failed to complete pending deletion for index
1> org.elasticsearch.env.ShardLockObtainFailedException: [idx][0]: thread interrupted while trying to obtain shard lock
1> 	at org.elasticsearch.env.NodeEnvironment$InternalShardLock.acquire(NodeEnvironment.java:866) ~[main/:?]
1> 	at org.elasticsearch.env.NodeEnvironment.shardLock(NodeEnvironment.java:775) ~[main/:?]
1> 	at org.elasticsearch.env.NodeEnvironment.lockAllForIndex(NodeEnvironment.java:718) ~[main/:?]
1> 	at org.elasticsearch.indices.IndicesService.processPendingDeletes(IndicesService.java:1071) ~[main/:?]
1> 	at org.elasticsearch.indices.cluster.IndicesClusterStateService$1.doRun(IndicesClusterStateService.java:410) [main/:?]
1> 	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:769) [main/:?]
1> 	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) [main/:?]
1> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
1> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
1> 	at java.lang.Thread.run(Thread.java:834) [?:?]
1> Caused by: java.lang.InterruptedException
1> 	at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedNanos(AbstractQueuedSynchronizer.java:1081) ~[?:?]
1> 	at java.util.concurrent.locks.AbstractQueuedSynchronizer.tryAcquireSharedNanos(AbstractQueuedSynchronizer.java:1369) ~[?:?]
1> 	at java.util.concurrent.Semaphore.tryAcquire(Semaphore.java:415) ~[?:?]
1> 	at org.elasticsearch.env.NodeEnvironment$InternalShardLock.acquire(NodeEnvironment.java:857) ~[main/:?]
1> 	... 9 more
1> [2019-08-26T00:20:00,956][INFO ][o.e.n.Node               ] [testUpdateSameHostSetting] stopping ...
```

This `InterruptedException` does seem to be unique to the failing case.

--- a/server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoverySourceService.java
+++ b/server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoverySourceService.java
@@ -24,10 +24,13 @@ import org.apache.logging.log4j.Logger;
 import org.elasticsearch.ExceptionsHelper;
 import org.elasticsearch.action.ActionListener;
 import org.elasticsearch.action.support.ChannelActionListener;
+import org.elasticsearch.action.support.PlainActionFuture;
 import org.elasticsearch.cluster.routing.ShardRouting;
 import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.component.AbstractLifecycleComponent;
 import org.elasticsearch.common.inject.Inject;
 import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.concurrent.FutureUtils;
 import org.elasticsearch.index.IndexService;
 import org.elasticsearch.index.shard.IndexEventListener;
 import org.elasticsearch.index.shard.IndexShard;
@@ -50,7 +53,7 @@ import java.util.Set;
  * The source recovery accepts recovery requests from other peer shards and start the recovery process from this
  * source shard to the target shard.
  */
-public class PeerRecoverySourceService implements IndexEventListener {
+public class PeerRecoverySourceService extends AbstractLifecycleComponent implements IndexEventListener {
 
     private static final Logger logger = LogManager.getLogger(PeerRecoverySourceService.class);
 
@@ -72,6 +75,19 @@ public class PeerRecoverySourceService implements IndexEventListener {
         this.recoverySettings = recoverySettings;
         transportService.registerRequestHandler(Actions.START_RECOVERY, ThreadPool.Names.GENERIC, StartRecoveryRequest::new,
             new StartRecoveryTransportRequestHandler());
+    }
+
+    @Override
+    protected void doStart() {
+    }
+
+    @Override
+    protected void doStop() {
+        ongoingRecoveries.awaitEmpty();
+    }
+
+    @Override
+    protected void doClose() {
     }
 
     @Override
@@ -118,6 +134,9 @@ public class PeerRecoverySourceService implements IndexEventListener {
     }
 
     final class OngoingRecoveries {
+
+        private final List<ActionListener<Void>> emptyListeners = new ArrayList<>();
+
         private final Map<IndexShard, ShardRecoveryContext> ongoingRecoveries = new HashMap<>();
 
         synchronized RecoverySourceHandler addNewRecovery(StartRecoveryRequest request, IndexShard shard) {
@@ -138,6 +157,10 @@ public class PeerRecoverySourceService implements IndexEventListener {
             if (shardRecoveryContext.recoveryHandlers.isEmpty()) {
                 ongoingRecoveries.remove(shard);
             }
+            if (ongoingRecoveries.isEmpty()) {
+                ActionListener.onResponse(emptyListeners, null);
+                emptyListeners.clear();
+            }
         }
 
         synchronized void cancel(IndexShard shard, String reason) {
@@ -155,6 +178,19 @@ public class PeerRecoverySourceService implements IndexEventListener {
                 }
                 ExceptionsHelper.maybeThrowRuntimeAndSuppress(failures);
             }
+        }
+
+        void awaitEmpty() {
+            assert lifecycle.stoppedOrClosed();
+            final PlainActionFuture<Void> future;
+            synchronized (this) {
+                if (ongoingRecoveries.isEmpty()) {
+                    return;
+                }
+                future = new PlainActionFuture<>();
+                emptyListeners.add(future);
+            }
+            FutureUtils.get(future);
         }
 
         private final class ShardRecoveryContext {

--- a/server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoverySourceService.java
+++ b/server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoverySourceService.java
@@ -135,9 +135,9 @@ public class PeerRecoverySourceService extends AbstractLifecycleComponent implem
 
     final class OngoingRecoveries {
 
-        private final List<ActionListener<Void>> emptyListeners = new ArrayList<>();
-
         private final Map<IndexShard, ShardRecoveryContext> ongoingRecoveries = new HashMap<>();
+
+        private List<ActionListener<Void>> emptyListeners = new ArrayList<>();
 
         synchronized RecoverySourceHandler addNewRecovery(StartRecoveryRequest request, IndexShard shard) {
             final ShardRecoveryContext shardContext = ongoingRecoveries.computeIfAbsent(shard, s -> new ShardRecoveryContext());
@@ -158,8 +158,11 @@ public class PeerRecoverySourceService extends AbstractLifecycleComponent implem
                 ongoingRecoveries.remove(shard);
             }
             if (ongoingRecoveries.isEmpty()) {
-                ActionListener.onResponse(emptyListeners, null);
-                emptyListeners.clear();
+                if (emptyListeners != null) {
+                    final List<ActionListener<Void>> onEmptyListeners = emptyListeners;
+                    emptyListeners = null;
+                    ActionListener.onResponse(onEmptyListeners, null);
+                }
             }
         }
 
@@ -188,6 +191,9 @@ public class PeerRecoverySourceService extends AbstractLifecycleComponent implem
                     return;
                 }
                 future = new PlainActionFuture<>();
+                if (emptyListeners == null) {
+                    emptyListeners = new ArrayList<>();
+                }
                 emptyListeners.add(future);
             }
             FutureUtils.get(future);

--- a/server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoverySourceService.java
+++ b/server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoverySourceService.java
@@ -137,9 +137,11 @@ public class PeerRecoverySourceService extends AbstractLifecycleComponent implem
 
         private final Map<IndexShard, ShardRecoveryContext> ongoingRecoveries = new HashMap<>();
 
-        private List<ActionListener<Void>> emptyListeners = new ArrayList<>();
+        @Nullable
+        private List<ActionListener<Void>> emptyListeners;
 
         synchronized RecoverySourceHandler addNewRecovery(StartRecoveryRequest request, IndexShard shard) {
+            assert lifecycle.started();
             final ShardRecoveryContext shardContext = ongoingRecoveries.computeIfAbsent(shard, s -> new ShardRecoveryContext());
             RecoverySourceHandler handler = shardContext.addNewRecovery(request, shard);
             shard.recoveryStats().incCurrentAsSource();

--- a/server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java
+++ b/server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java
@@ -402,13 +402,14 @@ public class RecoverySourceHandler {
         store.incRef();
         return Releasables.releaseOnce(() -> {
             final PlainActionFuture<Void> future = new PlainActionFuture<>();
-            threadPool.generic().execute(new ActionRunnable<>(future) {
-                @Override
-                protected void doRun() {
-                    store.decRef();
-                    listener.onResponse(null);
-                }
-            });
+            assert threadPool.generic().isShutdown() == false;
+            // TODO: We shouldn't use the generic thread pool here as we already execute this from the generic pool.
+            //       While practically unlikely at a min pool size of 128 we could technically block the whole pool by waiting on futures
+            //       below and thus make it impossible for the store release to execute which in turn would block the futures forever
+            threadPool.generic().execute(ActionRunnable.wrap(future, l -> {
+                store.decRef();
+                l.onResponse(null);
+            }));
             FutureUtils.get(future);
         });
     }

--- a/server/src/main/java/org/elasticsearch/node/Node.java
+++ b/server/src/main/java/org/elasticsearch/node/Node.java
@@ -592,6 +592,7 @@ public class Node implements Closeable {
                 .filter(p -> p instanceof LifecycleComponent)
                 .map(p -> (LifecycleComponent) p).collect(Collectors.toList());
             resourcesToClose.addAll(pluginLifecycleComponents);
+            resourcesToClose.add(injector.getInstance(PeerRecoverySourceService.class));
             this.pluginLifecycleComponents = Collections.unmodifiableList(pluginLifecycleComponents);
             client.initialize(injector.getInstance(new Key<Map<ActionType, TransportAction>>() {}), transportService.getTaskManager(),
                     () -> clusterService.localNode().getId(), transportService.getRemoteClusterService());
@@ -688,6 +689,7 @@ public class Node implements Closeable {
         assert localNodeFactory.getNode() != null;
         assert transportService.getLocalNode().equals(localNodeFactory.getNode())
             : "transportService has a different local node than the factory provided";
+        injector.getInstance(PeerRecoverySourceService.class).start();
         final MetaData onDiskMetadata;
         // we load the global state here (the persistent part of the cluster state stored on disk) to
         // pass it to the bootstrap checks to allow plugins to enforce certain preconditions based on the recovered state.
@@ -833,6 +835,7 @@ public class Node implements Closeable {
         toClose.add(injector.getInstance(IndicesService.class));
         // close filter/fielddata caches after indices
         toClose.add(injector.getInstance(IndicesStore.class));
+        toClose.add(injector.getInstance(PeerRecoverySourceService.class));
         toClose.add(() -> stopWatch.stop().start("cluster"));
         toClose.add(injector.getInstance(ClusterService.class));
         toClose.add(() -> stopWatch.stop().start("node_connections_service"));

--- a/server/src/test/java/org/elasticsearch/indices/recovery/PeerRecoverySourceServiceTests.java
+++ b/server/src/test/java/org/elasticsearch/indices/recovery/PeerRecoverySourceServiceTests.java
@@ -43,6 +43,7 @@ public class PeerRecoverySourceServiceTests extends IndexShardTestCase {
         StartRecoveryRequest startRecoveryRequest = new StartRecoveryRequest(primary.shardId(), randomAlphaOfLength(10),
             getFakeDiscoNode("source"), getFakeDiscoNode("target"), Store.MetadataSnapshot.EMPTY, randomBoolean(), randomLong(),
             SequenceNumbers.UNASSIGNED_SEQ_NO);
+        peerRecoverySourceService.start();
         RecoverySourceHandler handler = peerRecoverySourceService.ongoingRecoveries.addNewRecovery(startRecoveryRequest, primary);
         DelayRecoveryException delayRecoveryException = expectThrows(DelayRecoveryException.class,
             () -> peerRecoverySourceService.ongoingRecoveries.addNewRecovery(startRecoveryRequest, primary));

